test/testNodes.ts

```diff
@@ -0,0 +1,782 @@
+	return eqOk()
+}
+
+ASTCallArguments.prototype.isEqualNode = function isEqualNode(expected: ASTNode, path: TestPath)
+{
+	const next = nextPath(path, expected)
+	matchTypeAndToken(this, expected, next).unwrap()
+
+	const item = expected as ASTCallArguments
+	listEqual(this.arguments, item.arguments, 'arguments', next).unwrap()
+	return eqOk()
+}
+
+ASTTemplateArguments.prototype.isEqualNode = function isEqualNode(expected: ASTNode, path: TestPath)
+{
+	const next = nextPath(path, expected)
+	matchTypeAndToken(this, expected, next).unwrap()
+
+	const item = expected as ASTTemplateArguments
+	listEqual(this.arguments, item.arguments, 'arguments', next).unwrap()
+	return eqOk()
+}
+/* src/server/ast/values.ts end */
+
+/* src/server/ast/operations.ts start */
+ASTFunctionCall.prototype.isEqualNode = function isEqualNode(expected: ASTNode, path: TestPath)
+{
+	const next = nextPath(path, expected)
+	matchTypeAndToken(this, expected, next).unwrap()
+
+	const item = expected as ASTFunctionCall
+	this.args.isEqualNode(item.args, next).unwrap()
+	this.functionNameIdent.isEqualNode(item.functionNameIdent, next).unwrap()
+	undefinedEqual(this.templateArgs, item.templateArgs, 'templateArgs', next).unwrap()
+	return eqOk()
+}
+
+ASTUnaryOp.prototype.isEqualNode = function isEqualNode(expected: ASTNode, path: TestPath)
+{
+	const next = nextPath(path, expected)
+	matchTypeAndToken(this, expected, next).unwrap()
+
+	const item = expected as ASTUnaryOp
+	this.value.isEqualNode(item.value, next).unwrap()
+	return eqOk()
+}
+
+ASTBinaryOp.prototype.isEqualNode = function isEqualNode(expected: ASTNode, path: TestPath)
+{
+	const next = nextPath(path, expected)
+	matchTypeAndToken(this, expected, next).unwrap()
+
+	const item = expected as ASTBinaryOp
+	this.lhs.isEqualNode(item.lhs, next).unwrap()
+	this.rhs.isEqualNode(item.rhs, next).unwrap()
+	return eqOk()
+}
+
+ASTBetween.prototype.isEqualNode = function isEqualNode(expected: ASTNode, path: TestPath)
+{
+	const next = nextPath(path, expected)
+	matchTypeAndToken(this, expected, next).unwrap()
+
+	const item = expected as ASTBetween
+	this.lhs.isEqualNode(item.lhs, next).unwrap()
+	this.lop.isEqualNode(item.lop, next).unwrap()
+	this.value.isEqualNode(item.value, next).unwrap()
+	this.rop.isEqualNode(item.rop, next).unwrap()
+	this.rhs.isEqualNode(item.rhs, next).unwrap()
+	return eqOk()
+}
+
+ASTAssign.prototype.isEqualNode = function isEqualNode(expected: ASTNode, path: TestPath)
+{
+	const next = nextPath(path, expected)
+	matchTypeAndToken(this, expected, next).unwrap()
+
+	const item = expected as ASTAssign
+	this.ident.isEqualNode(item.ident, next).unwrap()
+	this.value.isEqualNode(item.value, next).unwrap()
+	return eqOk()
+}
+/* src/server/ast/operations.ts end */
+
+export function toBeEqualNode(received: ASTNode, expected: ASTNode)
+{
+	let res: TestPass
+
+	try
+	{
+		received.isEqualNode(expected, {path: 'root'}).unwrap()
+		res = {pass: true, message: () => ''}
```

Comment:
Rather than having a return result variable, is it possible to make this a direct value return, and same in the `catch` block?

Reply:
Directly returning the `{pass: true, ...}` object was my original plan but if we use `Result` and  `unwrap` in this hacky way, we can reduce a lot of boiler plate if statements.

For example with direct returns `ASTUnaryOp.prototype.isEqualNode` would look something like this
```ts
ASTUnaryOp.prototype.isEqualNode = function isEqualNode(expected: ASTNode, path: TestPath)
{
	const next = nextPath(path, expected)
	const match = matchTypeAndToken(this, expected, next)
	if (!match.pass)
		return match

	const item = expected as ASTUnaryOp
	const valueEq = this.value.isEqualNode(item.value, next)
	if (!valueEq.pass)
		return valueEq

	return eqOk()
}
```

But with unwrap it looks like this
```ts
ASTUnaryOp.prototype.isEqualNode = function isEqualNode(expected: ASTNode, path: TestPath)
{
	const next = nextPath(path, expected)
	matchTypeAndToken(this, expected, next).unwrap()

	const item = expected as ASTUnaryOp
	this.value.isEqualNode(item.value, next).unwrap()
	return eqOk()
}
```

This removes ~300 lines of if statments

Reply:
Right, but in this particular instance, we're saying does it not make more sense to do:
```ts
export function toBeEqualNode(received: ASTNode, expected: ASTNode) : TestPass
{
	try
	{
		received.isEqualNode(expected, {path: 'root'}).unwrap()
		return {pass: true, message: () => ''}
	}
	catch (err)
	{
		// Javascript erros cannot pass objects, only strings.
		// So we have to dig out the information from it in a really hacky way
		const msgLine = (err as Error).toString().split('\n')[0].split('Error: ')
		const json = JSON.parse(msgLine[msgLine.length - 1])
		return {pass: false, message: () => json.message}
	}
}
```
Given that TypeScript knows how to turn the {}'d expression into the type for the return

Reply:
Ooh I totally misunderstood that. That makes lot more sense :smile:. Will fix that too!